lib/ret/app_config.ex

```diff
@@ -0,0 +1,64 @@
+defmodule Ret.AppConfig do
+  use Cachex.Warmer
+  use Ecto.Schema
+  import Ecto.Changeset
+
+  alias Ret.{AppConfig, Repo}
+
+  @schema_prefix "ret0"
+  @primary_key {:app_config_id, :id, autogenerate: true}
+
+  schema "app_configs" do
+    field(:key, :string)
+    field(:value, :map)
+    belongs_to(:owned_file, Ret.OwnedFile, references: :owned_file_id)
+    timestamps()
+  end
+
+  def interval, do: :timer.seconds(15)
+
+  def execute(_state) do
+    {:ok, [{:app_config, get_config()}]}
+  end
+
+  def changeset(%AppConfig{} = app_config, attrs) do
+    # We wrap the config value in an outer %{value: ...} map because we want to be able to accept primitive
+    # value types, but store them as json.
+    attrs = attrs |> Map.put("value", %{value: attrs["value"]})
+
+    app_config
+    |> cast(attrs, [:key, :value])
+    |> unique_constraint(:key)
+  end
+
+  def get_config() do
+    AppConfig
+    |> Repo.all()
+    |> Map.new(fn app_config -> expand_key(app_config.key, app_config.value["value"], true) end)
+  end
+
+  defp expand_key(key, val, first) do
+    if key |> String.contains?("|") do
+      [head, tail] = key |> String.split("|", parts: 2)
```

Comment:
Can't use `.` or `_`, unfortunately, since I want to be able to handle keys like `"translations|en|home.hero_blurb"`. In toml, this is represented as `translations.en."home.hero_blurb"`, so it wouldn't be possible to use `.` without more intelligent parsing.

These transforms I'm doing are kinda unfortunate. I was wondering if I should just store the config as a single json blob record in the DB (or on disk?) instead. I don't have a strong opinion about it though.

Reply:
ah understood -- uniblob isn't a bad idea, but i think the current schema is better than a blob because it lets us fit in the stored file stuff nicely wrt referential integrity.